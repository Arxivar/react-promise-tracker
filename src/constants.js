/**
 * Emitter event types.
 * (Just one right now).
 */
export const PROGRESS_UPDATE = 'progress-update-event';

/**
 * Default values.
 */
export const DEFAULT_GROUP = 'default-group';
export const DEFAULT_DELAY = 0;
