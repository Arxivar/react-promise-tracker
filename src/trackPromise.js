import { Emitter } from './tinyEmitter';
import { PROGRESS_UPDATE, DEFAULT_GROUP } from './constants';
import { event } from './utils';

/**
 * Emitter singleton.
 */
export const emitter = new Emitter();

/**
 * Promise registry based on counters and its utility methods.
 */
const countMap = {};
const incrementCount = group => ++countMap[group] || ((countMap[group] = 1), 1);
const decrementCount = group => --countMap[group];
export const inProgress = group => {
  const counter = countMap[group || DEFAULT_GROUP];
  return Number.isInteger(counter) ? counter > 0 : undefined;
}

/**
 * Track a promise by increasing its group counter by one & programm its
 * untracking once the promise has settled.
 */
export const trackPromise = (promise, group) => {
  group = group || DEFAULT_GROUP;
  emitter.emit(event(PROGRESS_UPDATE, group), incrementCount(group) > 0);

  const onResolveHandler = () => untrack(group);
  promise.then(onResolveHandler, onResolveHandler);

  return promise;
};

/**
 * Untrack a promise by decrementing one to its group counter.
 */
const untrack = group => {
  emitter.emit(event(PROGRESS_UPDATE, group), decrementCount(group) > 0);
};

// TODO: Error management. Catch unhandled errors and throw events?
