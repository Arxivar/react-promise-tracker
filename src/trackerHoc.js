import React, { Component } from 'react'
import { emitter, promiseCounterUpdateEventId } from './trackPromise';
import { defaultArea } from './constants';

export const promiseTrackerHoc = (ComponentToWrap) => {
  return class promiseTrackerComponent extends Component {
    constructor(props) {
      super(props);

      this.state = {
        trackedPromiseInProgress: false,
        area: props.area || defaultArea,
      };
    }

    componentDidMount() {
      emitter.on(promiseCounterUpdateEventId, (anyPromiseInProgress, area) => {
        if (this.state.area === area) {
          this.setState({ trackedPromiseInProgress: anyPromiseInProgress });
        }
      });
    }

    componentWillUnmount() {
      emitter.off(promiseCounterUpdateEventId);
    }

    render() {
      return (
        <ComponentToWrap
          {...this.props}
          area={this.state.area}
          trackedPromiseInProgress={this.state.trackedPromiseInProgress}
        />
      )
    }
  }
}
