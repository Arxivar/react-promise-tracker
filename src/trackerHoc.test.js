import React from "react";
import { promiseTrackerHoc } from "./trackerHoc";
import * as trackPromiseAPI from "./trackPromise";
import { defaultArea } from "./constants";
import { act } from "react-dom/test-utils"; // ES6

describe("trackerHoc", () => {
  describe("Initial Status", () => {
    it('should render component with trackedPromiseInProgress equals false and area equals "default-area" when render promiseTrackerHoc without props', () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(<TrackedComponent />);

      expect(component).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals false, area equals "default-area" and customProp equals "test" when feeding customProp equals "test"', () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(<TrackedComponent customProp="test" />);

      expect(component).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals false and area equals "testArea" when feeding area equals "testArea"', () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(
        <TrackedComponent config={{ area: "testArea" }} />
      );

      expect(component).toMatchSnapshot();
    });

    it("should render component with trackedPromiseInProgress equals false when counter is 0", () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;
      trackPromiseAPI.getCounter = jest.fn().mockImplementation(() => 0);

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(<TrackedComponent />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(component).toMatchSnapshot();
    });

    it("should render component with trackedPromiseInProgress equals false when counter is 0 and emit event with progress equals false", () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;
      trackPromiseAPI.getCounter = jest.fn().mockImplementation(() => 0);

      const progress = false;
      const area = defaultArea;
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, "on")
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(<TrackedComponent />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(component).toMatchSnapshot();
    });

    it("should render component with trackedPromiseInProgress equals false when counter is 0 and emit event with progress equals false to different area", () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;
      trackPromiseAPI.getCounter = jest.fn().mockImplementation(() => 0);

      const progress = false;
      const area = "otherArea";
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, "on")
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(<TrackedComponent />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(component).toMatchSnapshot();
    });

    it("should render component with trackedPromiseInProgress equals true when counter is 0 and emit event with progress equals true", () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;
      trackPromiseAPI.getCounter = jest.fn().mockImplementation(() => 0);

      const progress = true;
      const area = defaultArea;
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, "on")
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(<TrackedComponent />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(component).toMatchSnapshot();
    });

    it("should render component with trackedPromiseInProgress equals false when counter is 0 and emit event with progress equals true to different area", () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;
      trackPromiseAPI.getCounter = jest.fn().mockImplementation(() => 0);

      const progress = true;
      const area = "otherArea";
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, "on")
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(<TrackedComponent />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(component).toMatchSnapshot();
    });

    it("should render component with trackedPromiseInProgress equals true when counter is 1", () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;
      trackPromiseAPI.getCounter = jest.fn().mockImplementation(() => 1);

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(<TrackedComponent />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(component).toMatchSnapshot();
    });

    it("should render component with trackedPromiseInProgress equals true when counter is 1 and emit event with progress equals true", () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;
      trackPromiseAPI.getCounter = jest.fn().mockImplementation(() => 1);

      const progress = true;
      const area = defaultArea;
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, "on")
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(<TrackedComponent />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(component).toMatchSnapshot();
    });

    it("should render component with trackedPromiseInProgress equals true when counter is 1 and emit event with progress equals true to different area", () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;
      trackPromiseAPI.getCounter = jest.fn().mockImplementation(() => 1);

      const progress = true;
      const area = "otherArea";
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, "on")
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(<TrackedComponent />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(component).toMatchSnapshot();
    });

    it("should render component with trackedPromiseInProgress equals false when counter is 1 and emit event with progress equals false", () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;
      trackPromiseAPI.getCounter = jest.fn().mockImplementation(() => 1);

      const progress = false;
      const area = defaultArea;
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, "on")
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(<TrackedComponent />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(component).toMatchSnapshot();
    });

    it("should render component with trackedPromiseInProgress equals true when counter is 1 and emit event with progress equals false to different area", () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;
      trackPromiseAPI.getCounter = jest.fn().mockImplementation(() => 1);

      const progress = false;
      const area = "otherArea";
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, "on")
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(<TrackedComponent />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(component).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals false and area equals "testArea" when feeding area equals "testArea" and delay equals 300', () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(
        <TrackedComponent config={{ area: "testArea", delay: 300 }} />
      );

      expect(component).toMatchSnapshot();
    });

    it("should render component with trackedPromiseInProgress equals false when counter is 0 and delay equals 300", () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;
      trackPromiseAPI.getCounter = jest.fn().mockImplementation(() => 0);

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(<TrackedComponent config={{ delay: 300 }} />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(component).toMatchSnapshot();
    });

    it("should render component with trackedPromiseInProgress equals false when counter is 0 and emit event with progress equals false and delay equals 300", () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;
      trackPromiseAPI.getCounter = jest.fn().mockImplementation(() => 0);

      const progress = false;
      const area = defaultArea;
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, "on")
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(<TrackedComponent config={{ delay: 300 }} />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(component).toMatchSnapshot();
    });

    it("should render component with trackedPromiseInProgress equals false when counter is 0 and emit event with progress equals false to different area and delay equals 300", () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;
      trackPromiseAPI.getCounter = jest.fn().mockImplementation(() => 0);

      const progress = false;
      const area = "otherArea";
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, "on")
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(<TrackedComponent config={{ delay: 300 }} />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(component).toMatchSnapshot();
    });

    it.only("should render component with trackedPromiseInProgress equals true when counter is 0 and emit event with progress equals true and delay equals 300", () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;
      trackPromiseAPI.getCounter = jest.fn().mockImplementation(() => 0);

      const progress = true;
      const area = defaultArea;
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, "on")
        .mockImplementation((id, callback) => callback(progress, area));
      const setTimeoutStub = jest
        .spyOn(window, "setTimeout")
        .mockImplementation(callback => callback());

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(<TrackedComponent config={{ delay: 300 }} />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(component).toMatchSnapshot();
    });

    it("should render component with trackedPromiseInProgress equals false when counter is 0 and emit event with progress equals true to different area and delay equals 300", () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;
      trackPromiseAPI.getCounter = jest.fn().mockImplementation(() => 0);

      const progress = true;
      const area = "otherArea";
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, "on")
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(<TrackedComponent config={{ delay: 300 }} />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(component).toMatchSnapshot();
    });

    it("should render component with trackedPromiseInProgress equals true when counter is 1 and delay equals 300", () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;
      trackPromiseAPI.getCounter = jest.fn().mockImplementation(() => 1);

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(<TrackedComponent config={{ delay: 300 }} />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(component).toMatchSnapshot();
    });

    it("should render component with trackedPromiseInProgress equals true when counter is 1 and emit event with progress equals true and delay equals 300", () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;
      trackPromiseAPI.getCounter = jest.fn().mockImplementation(() => 1);

      const progress = true;
      const area = defaultArea;
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, "on")
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(<TrackedComponent config={{ delay: 300 }} />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(component).toMatchSnapshot();
    });

    it("should render component with trackedPromiseInProgress equals true when counter is 1 and emit event with progress equals true to different area and delay equals 300", () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;
      trackPromiseAPI.getCounter = jest.fn().mockImplementation(() => 1);

      const progress = true;
      const area = "otherArea";
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, "on")
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(<TrackedComponent config={{ delay: 300 }} />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(component).toMatchSnapshot();
    });

    it("should render component with trackedPromiseInProgress equals false when counter is 1 and emit event with progress equals false and delay equals 300", () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;
      trackPromiseAPI.getCounter = jest.fn().mockImplementation(() => 1);

      const progress = false;
      const area = defaultArea;
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, "on")
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(<TrackedComponent config={{ delay: 300 }} />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(component).toMatchSnapshot();
    });

    it("should render component with trackedPromiseInProgress equals true when counter is 1 and emit event with progress equals false to different area and delay equals 300", () => {
      // Arrange
      const TestSpinnerComponent = props => <span>test</span>;
      trackPromiseAPI.getCounter = jest.fn().mockImplementation(() => 1);

      const progress = false;
      const area = "otherArea";
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, "on")
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const component = mount(<TrackedComponent config={{ delay: 300 }} />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(component).toMatchSnapshot();
    });
  });

  describe("Handling delay timeouts", () => {
    beforeEach(() => {
      jest.useFakeTimers();
    });

    afterAll(() => {
      jest.useRealTimers();
    });

    it.only("should render <h1>NO SPINNER</h2> when counter is 1 but delay is set to 200 (before timing out)", () => {
      // Arrange
      let TestSpinnerComponent = null;
      act(() => {
        TestSpinnerComponent = props => {
          return (
            <div>
              {props.promiseInProgress ? <h1>SPINNER</h1> : <h2>NO SPINNER</h2>}
            </div>
          );
        };

        trackPromiseAPI.getCounter = jest.fn().mockImplementation(() => 0);
      });

      // Act
      let component = null;
      act(() => {
        const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);
        component = mount(<TrackedComponent config={{ delay: 300 }} />);
      });

      // Check very beginning (no promises going on) NO SPINNER is shown
      // TODO: this assert could be skipped (move to another test)
      expect(component.text()).toMatch("NO SPINNER");
      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();

      // Assert
      // This promise will resolved after 1 seconds, by doing this
      // we will be able to test 2 scenarios:
      // [0] first 200ms spinner won't be shown (text NOSPINNER)
      // [1] after 200ms spinner will be shown (text SPINNER)
      // [2] after 1000ms spinner will be hidded again (text NOSPINNER)
      let myFakePromise = null;

      act(() => {
        myFakePromise = new Promise(resolve => {
          setTimeout(() => {
            resolve(true);
          }, 1000);
        });
      });

      act(() => {
        trackPromiseAPI.trackPromise(myFakePromise);

        // Runs all pending timers. whether it's a second from now or a year.
        // https://jestjs.io/docs/en/timer-mocks.html
        //jest.advanceTimersByTime(300);
      });

      act(() => {
        // Runs all pending timers. whether it's a second from now or a year.
        // https://jestjs.io/docs/en/timer-mocks.html
        jest.advanceTimersByTime(100);
      });

      // [0] first 200ms spinner won't be shown (text NOSPINNER)
      expect(component.text()).toMatch("NO SPINNER");

      act(() => {
        // Runs all pending timers. whether it's a second from now or a year.
        // https://jestjs.io/docs/en/timer-mocks.html
        jest.advanceTimersByTime(300);
      });

      // Before the promise get's resolved
      // [1] after 200ms spinner will be shown (text SPINNER)
      expect(component.text()).toMatch("SPINNER");

      // After the promise get's resolved

      act(() => {
        jest.runAllTimers();
      });

      // [2] after 1000ms spinner will be hidded again (text NOSPINNER)
      // Wait for fakePromise (simulated ajax call) to be completed
      // no spinner should be shown
      act(() => {
        myFakePromise.then(() => {
          expect(component.text()).toMatch("NO SPINNER");
        });
      });
    });
  });

});
