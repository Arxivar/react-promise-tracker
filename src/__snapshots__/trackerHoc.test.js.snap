// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trackerHoc should render component with trackedPromiseInProgress equals false and area equals "default" when render promiseTrackerHoc 1`] = `
<promiseTrackerComponent>
  <TestSpinnerComponent
    area="default"
    trackedPromiseInProgress={false}
  >
    <span>
      test
    </span>
  </TestSpinnerComponent>
</promiseTrackerComponent>
`;

exports[`trackerHoc should render component with trackedPromiseInProgress equals false and area equals "testArea" when feeding area equals "testArea" 1`] = `
<promiseTrackerComponent
  area="testArea"
>
  <TestSpinnerComponent
    area="testArea"
    trackedPromiseInProgress={false}
  >
    <span>
      test
    </span>
  </TestSpinnerComponent>
</promiseTrackerComponent>
`;

exports[`trackerHoc should render component with trackedPromiseInProgress equals false, area equals "default" and customProp equals "test" when feeding customProp equals "test" 1`] = `
<promiseTrackerComponent
  customProp="test"
>
  <TestSpinnerComponent
    area="default"
    customProp="test"
    trackedPromiseInProgress={false}
  >
    <span>
      test
    </span>
  </TestSpinnerComponent>
</promiseTrackerComponent>
`;
